modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPartitionsExchangeFuture.java

```diff
@@ -1225,6 +1232,82 @@ private void distributedExchange() throws IgniteCheckedException {
         initDone();
     }
 
+    /**
+     *
+     */
+    private void onDistributedExchangeTimeout() {
+        if (!isDone()) {
+            synchronized (mux) {
+                if (crd.isLocal()) {
+                    Set<UUID> remaining0 = new HashSet<>(this.remaining);
+
+                    log.warning("Exchange timeout reached on coordinator node, kicking pending nodes: " +
+                        S.arrayToString(UUID.class, remaining0.toArray()));
+
+                    for (UUID remainingUuid : remaining0)
+                        cctx.discovery().failNode(remainingUuid, null);
+                }
+                else {
+                    try {
+                        log.warning("Exchange timeout reached on non-coordinator node, " +
+                            "sending check request to coordinator");
+
+                        cctx.kernalContext().io().sendToGridTopic(
+                            crd, GridTopic.TOPIC_EXCHANGE, new PartitionsExchangeFinishedCheckRequest(), SYSTEM_POOL);
+
+                        cctx.time().schedule(() -> {
+                            if (!isDone())
+                                cctx.kernalContext().failure().process(
+                                    new FailureContext(FailureType.CRITICAL_ERROR,
+                                        new TimeoutException("Last exchange check response wasn't received.")));
+                        }, cctx.gridConfig().getExchangeHardTimeout(), -1);
+                    }
+                    catch (IgniteCheckedException ex) {
+                        cctx.kernalContext().failure().process(new FailureContext(FailureType.CRITICAL_ERROR, ex));
+
+                        throw new IgniteException(ex);
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * @param lastFinishedVer Last finished topology version.
+     * @param pendingVer Pending topology version.
+     * @param receivedSingleMsg Received single message flag.
+     */
+    public void onCrdLastFinishedVersionReceived(AffinityTopologyVersion lastFinishedVer,
+        AffinityTopologyVersion pendingVer, boolean receivedSingleMsg) {
+        log.warning("Received check response from coordinator: " +
+            "[lastFinishedVer=" + lastFinishedVer + ", pendingVer=" + pendingVer +
+            ", receivedSingleMsg=" + receivedSingleMsg + ", localExchId=" + exchId + "]");
+
+        boolean failLocal;
+
+        if (lastFinishedVer.compareTo(exchId.topologyVersion()) >= 0)
+            failLocal = true;
+        else if (pendingVer.equals(exchId.topologyVersion()) && !receivedSingleMsg)
+            failLocal = true;
+        else if (pendingVer.equals(exchId.topologyVersion()) && receivedSingleMsg)
+            return; // Do nothing, this case should be resolved by coordinator.
+        else
+            failLocal = false;
```

Comment:
Consider a situation, when a non-coordinator node didn't receive one of the messages, that change topology version, due to a bug in discovery. Let's say, coordinator has pending exchange with `topVer=5` and the remaining node – `topVer=6`. In this case the remaining node will decide, that coordinator should die. After coordinator re-election this situation may repeat until the problematic node is left the only one in the cluster.